eth/handler.go

```diff
@@ -338,6 +335,30 @@ func (pm *ProtocolManager) handle(p *peer) error {
 		p.Log().Info("Ethereum handshake failed", "err", err)
 		return err
 	}
+	forcePeer := false
+	if handler, ok := pm.engine.(consensus.Handler); ok {
+		isValidator, err := handler.Handshake(p)
+		if err != nil {
+			p.Log().Warn("Istanbul handshake failed", "err", err)
+			return err
+		}
+		forcePeer = (isValidator == true)
+		p.Log().Trace("Peer completed Istanbul handshake", "forcePeer", forcePeer)
+	}
+	// Ignore max peer and max inbound peer check if:
+	//  - this is a trusted or statically dialed peer
+	//  - the peer is from from the proxy server (e.g. peers connected to this node's internal network interface)
+	//  - forcePeer is true
+	if !forcePeer {
+		if err := p.Peer.Server.CheckPeerCounts(p.Peer); err != nil {
+			return err
+		}
+		isStaticOrTrusted := p.Peer.Info().Network.Trusted || p.Peer.Info().Network.Static
+		if !isStaticOrTrusted && pm.peers.Len() >= pm.maxPeers && p.Peer.Server != pm.proxyServer {
```

Comment:
How is this check different than the one done by `CheckPeerCounts`? If this is correct, could you add a comment to highlight that?